integration/framework/service.go

```diff
@@ -0,0 +1,345 @@
+package framework
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/expfmt"
+)
+
+var (
+	dockerPortPattern = regexp.MustCompile(`^.*:(\d+)$`)
+)
+
+type Service struct {
+	name         string
+	image        string
+	networkName  string
+	networkPorts []int
+	env          map[string]string
+	command      *Command
+	readiness    *ReadinessProbe
+
+	// Maps container ports to dynamically binded local ports.
+	networkPortsContainerToLocal map[int]int
+
+	// Generic retry backoff.
+	retryBackoff *util.Backoff
+}
+
+func NewService(
+	name string,
+	image string,
+	networkName string,
+	networkPorts []int,
+	env map[string]string,
+	command *Command,
+	readiness *ReadinessProbe,
+) *Service {
+	return &Service{
+		name:                         name,
+		image:                        image,
+		networkName:                  networkName,
+		networkPorts:                 networkPorts,
+		env:                          env,
+		command:                      command,
+		networkPortsContainerToLocal: map[int]int{},
+		readiness:                    readiness,
+		retryBackoff: util.NewBackoff(context.Background(), util.BackoffConfig{
+			MinBackoff: 100 * time.Millisecond,
+			MaxBackoff: 200 * time.Millisecond,
+			MaxRetries: 100, // Sometimes the CI is slow
+		}),
+	}
+}
+
+func (s *Service) Start() (err error) {
+	// In case of any error, if the container was already created, we
+	// have to cleanup removing it. We ignore the error of the "docker rm"
+	// because we don't know if the container was created or not.
+	defer func() {
+		if err != nil {
+			_, _ = RunCommandAndGetOutput("docker", "rm", "--force", s.name)
+		}
+	}()
+
+	cmd := exec.Command("docker", s.buildDockerRunArgs()...)
+	cmd.Stdout = &LinePrefixWriter{prefix: s.name + ": ", wrapped: os.Stdout}
+	cmd.Stderr = &LinePrefixWriter{prefix: s.name + ": ", wrapped: os.Stderr}
+	if err = cmd.Start(); err != nil {
```

Comment:
This does not seem to detect some failures:
```
Starting table-manager
table-manager: Unable to find image 'cortex:latest' locally
table-manager: docker: Error response from daemon: pull access denied for cortex, repository does not exist or may require 'docker login'.
table-manager: See 'docker run --help'.
panic: test timed out after 5m0s
```

Reply:
Thanks for spotting it! There was a bug in `WaitStarted()`. I was calling `backoff.Reset()` instead of `backoff.Wait()` 🤦‍♂